### PR TITLE
Fix 21950: Rewrite function calls depending on noreturn arguments 

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
Replaces the function call with the evaluation of the arguments up to
the first `noreturn` parameter. The remaining arguments / the function
call will never be executed anyways and the backend cannot handle
`noreturn` parameters.

---

Also added a bunch of test not strictly related to this change. Some of them are currently disabled because they require further changes - but I think it's worthwhile to upstream them as is.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by kinke** [comment]
@MoonlightSentinel: If such things aren't handled in the frontend, each glue layer must replicate this ugly stuff.

**Comment by MoonlightSentinel** [comment]
It's a tradeoff between more destructive rewriting done by the frontend. I can move this into the semantic if that's the consensus for all compiler (CC @ibuclaw)

TBH, didn't expect problems for LLVM to deal with `noreturn` parameters. Rewriting the call was the second best solution because the backend chokes on parameters and trying to fix that was ... quite impossible given the state of the code.

**Comment by kinke** [comment]
Thinking some more about it - is this acually correct?
```D
noreturn loop() { while (true) {} }
noreturn bla(int i, noreturn);
void main() { bla(1+2, loop()); }
```
Shouldn't `loop()` be actually called in this example?

**Comment by MoonlightSentinel** [comment]
> speaking of which, what does `T foo(int a, noreturn b)` mean? Can such a function ever be called or is this a 1337 way of encoding `cast(void) a`?

See the example in [the DIP](https://github.com/dlang/DIPs/blob/master/DIPs/accepted/DIP1034.md). Both parameters will be evaluated but the function won't be called because the expression yielding the second argument will throw / exit / loop indefinitely.

**Comment by MoonlightSentinel** [comment]
> Shouldn't loop() be actually called in this example?

Yes, this rewrite should only omit the arguments *after* `loop()` (if any)

**Comment by kinke** [comment]
Yeah sorry just tested it now with a fresh DMD build, works fine, it includes the first `noreturn` one, missed that.

The rewrite during semantic would allow eliding emission/declaration of these ghost functions in the glue layer.

File: test/runnable/noreturn2.d [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        noreturn a;
+        noreturn b = cast(noreturn) 1;
+    });
+
+    if (false) // Read does not assert
+    testAssertFailure(__LINE__ + 3, msg, function noreturn()
+    {
+        noreturn a;
+        return a;
+    });
```
Comment (ibuclaw) [inline test/runnable/noreturn2.d:138]: Isn't this test broken?  A good compiler would complain that this noreturn function infact returns.

Comment (MoonlightSentinel) [reply test/runnable/noreturn2.d:138]: It doesn't return, reading `a` triggers an `assert(0)` as specified under [Declarations](https://github.com/dlang/DIPs/blob/master/DIPs/accepted/DIP1034.md#interaction-with-other-language-features)

**Pull request merged**